Added AODNet Image Dehazing Example

soumik12345 opened the pull request:
Image Dehazing is an important problem in Computer Vision. It can not only be used to enhance photographs clicked in hazy conditions, but also can be used to augment other computer vision systems such as Object Detection and Semantic Segmentation as a pre-processing step. In this example, we implement [AODNet](https://arxiv.org/abs/1707.06543v1), a simple and lightweight Deep CNN model that directly generates a clean image from a hazy image. The end-to-end design of AODNet makes it easy to embed AOD-Net into other deep models such as Faster-RCNN and YOLO as a pre-processing step, making it easier for such a model to detect and recognize objects from a clean image.

soumik12345 commented:
@fchollet I would continue to refactor the code for Image Dehazing as per your suggestions in [https://github.com/keras-team/keras-io/pull/440](https://github.com/keras-team/keras-io/pull/440) since I accidentally deleted the fork of this [https://github.com/keras-team/keras-io/pull/439](https://github.com/keras-team/keras-io/pull/439).

fchollet commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,367 @@
+import numpy as np
+from PIL import Image
+from glob import glob
+from glob import glob
```
Please apply fixes to all comments I made on the previous PR?

soumik12345 commented:
- [x] I kept the `!nvidia-smi -L`, I think it would be better for reproducibility if we know which GPU the demo was trained on.
- [x] Removed redundant import line.
- [x] Refactored markdown links on a single line.
- [x] Replaced `os.system` with shell commands.
- [ ] Replace `get_image_file_list` using `image_dataset_from_directory`.
- [ ] Replace augmentations with preprocessing layer
- [x] Replaced the dataloader class with a stateless function.
- [x] Removed specifications of default parameters in the model.
- [x] Replaced `tf.initializers` with corresponding `tf.keras.initializers`.
- [x] Replaced `tf.math.multiply(k, inputs)` with `k * inputs`.
- [x] Removed unnecessary use of `OneDeviceStrategy`.

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+"""
+# Introduction
+
+Image Dehazing is an important problem is Computer Vision. It can not only be used to
+enhance photographs clicked in hazy conditions, but also can be used to augment other
+computer vision systems such as Object Detection and Seantic Segmentation as a
```
Suggestion:

- Use sentence case (insteda, we could add some kind of _emphasis_ on non-commonly capitalized words)
- Fix repetition ("can be used to" x 2)

```suggestion
_Image dehazing_ is an important problem is computer vision that can be used not 
only to enhance photographs clicked in hazy conditions, but also can be used to
augment other computer vision systems, such as _object detection_ and _seantic
segmentation as a
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+enhance photographs clicked in hazy conditions, but also can be used to augment other
+computer vision systems such as Object Detection and Seantic Segmentation as a
+pre-processing step. In this example, we implement
+[AODNet](https://arxiv.org/abs/1707.06543v1), a simple and light-weight Deep CNN model
+that directly generates a clean image from a hazy image. The end-to-end design of AODNet
```
Suggestion:
- Clarify the abbreviation ("All-in-One Network (AOD-Net)")
- Use "AOD-Net" with a dash (as per https://arxiv.org/abs/1707.06543v1)
- Use a dash to provide a defition
- "Can generate X directly from Y" could be better
- "...deep CNN" - non capitalized
- "the AOD-Net" with the definite article ("the ... network")

```suggestion
the [All-in-One Network (AODNet)](https://arxiv.org/abs/1707.06543v1) — a simple and
lightweight deep convolutional neural network model that can generate a clean
image directly from a hazy image. The end-to-end design of the AOD-Net
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+pre-processing step. In this example, we implement
+[AODNet](https://arxiv.org/abs/1707.06543v1), a simple and light-weight Deep CNN model
+that directly generates a clean image from a hazy image. The end-to-end design of AODNet
+makes it easy to embed AOD-Net into other deep models such as Faster-RCNN and YOLO as a
+pre-processing step, making it easier for such model to detect and recognize objects from
+a clean image.
```
Suggestion:

- Provide links to papers (YOLO, Faster-RCNN)
- Fix repetition ("make it easy" x 2) with "enable"

```suggestion
enables embedding the network into other neural network models for, such as
[Faster-RCNN](https://arxiv.org/abs/1506.01497v3) and
[YOLO](https://pjreddie.com/media/files/papers/yolo.pdf), as a pre-processing
step, making it easier for such model to detect and recognize objects from
clean images.
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+Author: [Soumik Rakshit](https://github.com/soumik12345)
+Date created: 2021/04/10
+Last modified: 2021/04/10
+Description: Train a lightweight Deep CNN for image denhazing.
```
Suggestion:

- Use sentence case, refactor

```suggestion
Description: Dehaze images with a lightweight deep convolutional neural network.
```

("AOD-Net" is defined later in the intro)

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+import matplotlib.pyplot as plt
+
+"""
+# Downloading Dataset
```
Suggestion:

```suggestion
# Load the NYU2 Depth Database
```

(similar to https://keras.io/examples/vision/image_classification_from_scratch/)

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+"""
+# Downloading Dataset
+
+The authors of the AODNet paper have created synthesized hazy images using the
+groundtruth images with depth meta-data from the indoor 
+[NYU2 Depth Database](https://cs.nyu.edu/~silberman/datasets/nyu_depth_v2.html). We will
+download this dataset from Google Drive using `gdown`.
```
Suggestion:

```suggestion
The experiment in the AOD-Net paper used synthesized hazy images from the
ground-truth images with depth meta-data from the indoor 
[NYU2 Depth Database](https://cs.nyu.edu/~silberman/datasets/nyu_depth_v2.html).
We will download this dataset from Google Drive using `gdown`.
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+"""
+
+"""
+## Data Loader
```
Suggestion:

```suggestion
## Define the data transformation and augmentation functions
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+print(val_dataset)
+
+"""
+## Building AODNet Model
+
+Now we will build AODNet a subclass of  `tf.keras.Model`. The following diagrams taken
+from the paper summarizes the architecture of the AODNet model.
```
Suggestion:

- Nit (grammar, etc)

```suggestion
## Define the AOD-Net model

Now we will define the AOD-Net model as a subclass of  `tf.keras.Model`. The
following diagrams from the original paper summarize its architecture:
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+Now we will build AODNet a subclass of  `tf.keras.Model`. The following diagrams taken
+from the paper summarizes the architecture of the AODNet model.
+
+### The AODNet Model
```
```suggestion
### The AOD-Net model
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+### The AODNet Model
+![](https://i.imgur.com/nmnF0cY.png)
+
+### K-estimation Model
```
```suggestion
### The K-estimation model
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+
+
+"""
+Since we are training the model using a single GPU, i.e, Nvidia Tesla P100, we would use
+`tf.distribute.OneDeviceStrategy`, using this strategy will place any variables created
+in its scope on the specified device. We would define and compile our model in the scope
+of our distribution strategy. We would be using Adam as out optimizer, Mean Squared Error
+as the loss function and PSNR (Peak Signal Noise Ratio) as our metric.
+
+**Note:** For training AODNet in a multi-GPU environment,
+`tf.distribute.MirroredStrategy` can be used as a distribution stratgey.
```
Suggestion:

- Add emphasis to "TensorFlow distribution strategy" (with this guide x 2: https://www.tensorflow.org/guide/distributed_training)
- Use the definite article "the"
- Fix a typo ("strategy")

```suggestion
In this example, we are training the model using a single GPU (an NVIDIA Tesla
P100) and the`tf.distribute.OneDeviceStrategy`. This
[TensorFlow distribution
strategy](https://www.tensorflow.org/tutorials/distribute/keras)
will place any variables created in its scope on the specified device. We will
define and compile our model in the scope of our distribution strategy, using
Adam as the optimizer, the Mean Squared Error
as the loss function, and the PSNR (Peak Signal Noise Ratio) as our metric.

**Note:** For training the AOD-Net in a multi-GPU environment, we can also use
`tf.distribute.MirroredStrategy` (go to the
[Distributed training](https://www.tensorflow.org/guide/distributed_training)
guide more information).
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+"""
+# Training
+
+We train AODNet for 10 epochs, on Nvidia Tesla P100, an epoch takes around 190 seconds,
+so the model takes around 30 minutes to train
```
```suggestion
We train the AOD-Net for 10 epochs. On an NVIDIA Tesla P100, each epoch can take around 190 seconds, so the model may takes around 30 minutes to train.
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,354 @@
+"""
+# Inference
+
+We define 2 simple utility functions for inferring from an image and plotting the results
```
```suggestion
Next, let's define 2 simple utility functions for inferring from an image and plotting 
the results:
```

8bitmp3 reviewed (commented):
Another awesome example @soumik12345.

I left a few suggestions, PTAL 👍 hope they help.

soumik12345 commented:
> Another awesome example @soumik12345.
> 
> I left a few suggestions, PTAL 👍 hope they help.

@8bitmp3 Thanks for all the suggestions, updated the example as per your suggestions :)

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,362 @@
+# Introduction
+
+_Image dehazing_ — improving the image quality by removing the haze from images
+captured in real-world weather conditions — is an important problem is computer
```
Oops, a typo!

```suggestion
captured in real-world weather conditions — is an important problem in computer
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,362 @@
+the [All-in-One Network (AOD-Net)](https://arxiv.org/abs/1707.06543v1) — a simple and
+lightweight deep convolutional neural network model that can generate clean
+images directly from hazy images. The end-to-end design of the AOD-Net
+enables embedding the network into other neural network models for, such as
```
```suggestion
enables embedding the network into other neural network models, such as
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,362 @@
+"""
+## Define the AOD-Net model
+
+Now we will define the AOD-Net model as a subclass of  `tf.keras.Model`. The
```
Nit:

```suggestion
Now we will define the AOD-Net model as a subclass of `tf.keras.Model`. The
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,362 @@
+## Define the AOD-Net model
+
+Now we will define the AOD-Net model as a subclass of  `tf.keras.Model`. The
+following diagrams from the original paper summarize its architecture:
```
Suggestion: "its" -> "the model's" (to make sure some users don't get confused)

```suggestion
following diagrams from the original paper summarize the model's architecture:
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,362 @@
+training_history = model.fit(train_dataset, validation_data=val_dataset, epochs=10)
+
+"""
+## Plot of Training and Validation Loss
```
Suggestion: use sentence case, "losses" (plural)

```suggestion
## Plot of training and validation losses
```

8bitmp3 commented on examples/vision/aodnet.py:
```diff
@@ -0,0 +1,362 @@
+plt.show()
+
+"""
+## Plot of Training and Validation PSNR
```
Suggestion: use sentence case, "PSNRs" (since it is "ratios", it should be plural)

```suggestion
## Plot of training and validation PSNRs
```

fchollet commented:
Please let me know when the PR is ready for review again! :)

soumik12345 commented:
> Please let me know when the PR is ready for review again! :)

Sure @fchollet

soumik12345 commented:
@fchollet I attempted to rewrite the augmentations using Keras preprocessing layers, but the problem that I've been unable to solve is the application of exactly the same transform to both the cropped segment of the input image and the corresponding ground truth. Would really appreciate your thoughts on this topic.

fchollet commented:
> I attempted to rewrite the augmentations using Keras preprocessing layers, but the problem that I've been unable to solve is the application of exactly the same transform to both the cropped segment of the input image and the corresponding ground truth. Would really appreciate your thoughts on this topic.

I would recommend creating 2 identical layers (or sets of layers) with the same `seed` argument in the constructor. They will apply the same transformations as long as they're called in the same order.

fchollet commented:
Are you still working on this?

fchollet commented:
Closing PR as stale.
